### PR TITLE
Add new Lint/ConstantDefinitionInBlock cop

### PR DESCRIPTION
**Pull request by eugeneius**
Constants defined within a block are added to the current lexical scope, which means the block has no effect on where the constant is defined or how it can be accessed. I often see constants defined inside Rake tasks or RSpec `describe` blocks where the intent was clearly for them to be scoped to the block, but which actually leak into the top-level scope.

-----------------

Before submitting the PR make sure the following are checked:

* [x] Wrote [good commit messages][1].
* [x] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Added an entry to the [Changelog](https://github.com/rubocop-hq/rubocop/blob/master/CHANGELOG.md) if the new code introduces user-observable changes. See [changelog entry format](https://github.com/rubocop-hq/rubocop/blob/master/CONTRIBUTING.md#changelog-entry-format).
* [x] The PR relates to *only* one subject with a clear title and description in grammatically correct, complete sentences.
* [x] Run `bundle exec rake default`. It executes all tests and RuboCop for itself, and generates the documentation.

[1]: https://chris.beams.io/posts/git-commit/

### CODE REVIEW

File: lib/rubocop/cop/lint/constant_defined_in_block.rb [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+      #   # good
+      #   task :lint do
+      #     files_to_lint = Dir['lib/*.rb']
+      #   end
+      #
+      #   # good
+      #   describe 'making a request' do
+      #     let(:test_request) { Class.new }
+      #   end
+      class ConstantDefinedInBlock < Base
```
Comment (bbatsov) [inline lib/rubocop/cop/lint/constant_defined_in_block.rb:32]: Probably we should name this `ConstantDefinitionInBlock` to be more consistent with existing cop names.

**Comment by bbatsov** [comment]
Nice cop! I think we should also mention this in the Ruby Style Guide, as it might not be super obvious to some people.

File: lib/rubocop/cop/lint/constant_defined_in_block.rb (hunk 1 above)
Comment (eugeneius) [reply lib/rubocop/cop/lint/constant_defined_in_block.rb]: Good call, I updated the name.

**Comment by eugeneius** [comment]
I opened https://github.com/rubocop-hq/ruby-style-guide/pull/841 to add a corresponding guideline to the Ruby Style Guide.

**Comment by bbatsov** [comment]
Thanks! I'll merge the cop as it is, as I might cut a new release today, but we should add a link to the new guideline to the cop metadata.

**Comment by ghiculescu** [comment]
FYI for anyone who comes across this, this cop isn't compatible with `sorbet`. Specifically, constants defined in blocks are [how enums work](https://sorbet.org/docs/tenum).

@marcandre do you want me to do a docs PR for this one too?

**Comment by marcandre** [comment]
> @marcandre do you want me to do a docs PR for this one too?

Yes, that would be a great starting point. At a minimum we should document these cases, and afterwards see what is the best to address them. In this case, we could exclude `enum do`, or have an allowlist, for example...

**Pull request merged**